### ISSUE TITLE
No IP address in request object

### ISSUE THREAD

**Issue by libersoft-org** [elysiajs/elysia#325]
I'd like to store client's IP addresses, but it seems like there is no IP parameter in request, e.g.:

const app = new Elysia().get('/*', (req) => console.log(req));

Returns:

```
{
  request: Request (0 KB) {
    method: "GET",
    url: "http://shortener/",
    headers: Headers {
      "host": "shortener",
      "connection": "close",
      "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/119.0",
      "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
      "accept-language": "cs,sk;q=0.8,en-US;q=0.5,en;q=0.3",
      "upgrade-insecure-requests": "1",
      "sec-fetch-dest": "document",
      "sec-fetch-mode": "navigate",
      "pragma": "no-cache",
      "cache-control": "no-cache",
      "accept-encoding": "gzip, deflate, br",
      "sec-fetch-site": "none",
      "sec-fetch-user": "?1"
    }
  },
  store: {},
  set: {
    cookie: {},
    headers: {},
    status: 200
  },
  qi: -1,
  path: "/",
  params: {
    "*": ""
  },
  headers: {
    host: "shortener",
    connection: "close",
    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/119.0",
    accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
    "accept-language": "cs,sk;q=0.8,en-US;q=0.5,en;q=0.3",
    "upgrade-insecure-requests": "1",
    "sec-fetch-dest": "document",
    "sec-fetch-mode": "navigate",
    pragma: "no-cache",
    "cache-control": "no-cache",
    "accept-encoding": "gzip, deflate, br",
    "sec-fetch-site": "none",
    "sec-fetch-user": "?1"
  },
  cookie: {},
  query: {}
}
```

No client's IP mentioned. :(

**Comment by Jatin-MYST** [comment]
if you are using a reverse proxy you can set client address to a header like this.
nginx example:
proxy_set_header X-Real-IP $remote_addr;

Then look for X-Real-IP in bun request headers.

**Comment by libersoft-org** [comment]
> if you are using a reverse proxy you can set client address to a header like this. nginx example: proxy_set_header X-Real-IP $remote_addr;
> 
> Then look for X-Real-IP in bun request headers.

Thank you for your answer. Well, this works only on reverse proxy, but what about on standalone Elysia / Bun server?

**Comment by SaltyAom** [comment]
See https://www.npmjs.com/package/elysia-ip

**Comment by libersoft-org** [comment]
> See https://www.npmjs.com/package/elysia-ip

Yes, I saw this too, but it does not work at all. in ".use(ip())" part it throws an error: "null is not an object (evaluating 'app.server.requestIP')"

**Comment by paperdave** [comment]
i implemented requestIP in bun 1.0.4. please update if you are below that version

elysia-ip works for me but it seems they allow a ton of headers as the ip so by default it is vulnerable to spoofing. here is that package in an inline plugin using only server.requestIP

```ts
import { Elysia } from "elysia";

new Elysia()
  .use(app => app.derive(({ request }) => ({ ip: app.server?.requestIP(request) })))
  .get("/", ({ ip }) => ip).listen(3000);
```

**Comment by gaurishhs** [comment]
hi! `elysia-ip` developer here. 

> ".use(ip())" part it throws an error: "null is not an object (evaluating 'app.server.requestIP')"

there are 2 possible reasons for this:

1. Either the server hasn't started yet or for some reasons `Elysia.server` is undefined. 
2. Using multiple Elysia instances, Check https://github.com/elysiajs/elysia/issues/230#issuecomment-1748647185 It is probably how plugins inside of Elysia instances work because the plugin is nothing other than the code sample provided by dave except that works for only Bun and i had to add support for other runtimes too in the package since Elysia does work on multiple runtimes. 

> but it seems they allow a ton of headers as the ip so by default it is vulnerable to spoofing

If you're using latest version of the library along with Bun, The library does not even check for headers and returns the ip using early returns. 
If you're using some another Runtime, It checks for header and the highest priority goes to `X-Forwarded-For` which is the de-factor standard header by [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) which is present 99% of the times 

After all this, The library checks for other headers used by Reverse proxies and servers. You can even change them by specifying the `checkHeaders` option when calling the ip function 

```ts
.use(ip({ checkHeaders: ['xyz'] })
```

When checkHeaders is specified, `X-Forwarded-For` is ignored.

**Comment by libersoft-org** [comment]
> i implemented requestIP in bun 1.0.4. please update if you are below that version
> 
> elysia-ip works for me but it seems they allow a ton of headers as the ip so by default it is vulnerable to spoofing. here is that package in an inline plugin using only server.requestIP
> 
> ```ts
> import { Elysia } from "elysia";
> 
> new Elysia()
>   .use(app => app.derive(({ request }) => ({ ip: app.server?.requestIP(request) })))
>   .get("/", ({ ip }) => ip).listen(3000);
> ```

This returns undefined.

**Comment by gaurishhs** [comment]
I think the issue is not using app.listen if you're using Bun you should probably be using app.listen instead of Bun.serve

**Comment by paperdave** [comment]
> The library does not even check for headers and returns the ip using early returns

ah i misread the early `return`

**Comment by libersoft-org** [comment]
> I think the issue is not using app.listen if you're using Bun you should probably be using app.listen instead of Bun.serve

I cannot use .listen as it does not work with Unix sockets. But I think this won't be the problem.
